New error extension `http` for setting status/headers in context, resolvers (#6857)
apollographql/apollo-server: docs/source/data/errors.mdx

The REDACTED doesn't have sufficient privileges.
```

## Setting HTTP status code and headers

GraphQL, by design, does not use the same conventions from REST to communicate via HTTP verbs and status codes. Client information should be contained in the schema or as part of the standard response `errors` field. We recommend using the included [Error Codes](#built-in-error-codes) or [Custom Errors](#custom-errors) for error consistency rather than directly modifying the HTTP response.

> Apollo Server's general behavior with respect to status codes is as follows:

Reviewer: ```suggestion
Apollo Server uses different HTTP status codes in various situations:
```

> - If Apollo Server has not properly started up or is in the process of shutting down, use status code 500. (The former only happens if you're using a serverless integration instead of `await`ing `startStandaloneServer(server)` or `server.start()`; the latter only happens if you have not properly set up draining for your non-standalone integration.)

Reviewer: I feel like you can remove the second sentence here; I think the first sentence is clear enough that the extra background isn't necessary. 
```suggestion
- If Apollo Server hasn't properly started up or is in the process of shutting down, it responds with a 500 status code.
```

Author: My point here is that if you're seeing this it's because you've set up your server in a non-optimal way (which is easy to do because it does take a bit more work in anything other than startStandaloneServer to do it right) — so if people come here concerned their server is sending 500s it would be nice to tell them there's a better way. Maybe could be expressed better though?

Reviewer: Okay, I see what you mean. 

Could you further clarify what "if you're using a serverless integration instead of `await`ing `startStandaloneServer(server)` or `server.start()`" means here? Does that mean this error code will only happen with serverless integrations where the server doesn't start up properly? Or if you forget to `await` your server's startup with non-serverless integrations? Or both? 

I think we should we could make this a sub-bullet point, so something like (this suggestion will need to be tweaked based on the answer to my question above):

```suggestion
-  If Apollo Server hasn't properly started up or is in the process of shutting down, it responds with a 500 status code. 
  - The former happens if you haven't `await`ed your server's start-up, and the latter if you aren't properly [draining your server](/apollo-server/api/plugin/drain-http-server/#using-the-plugin).
```

Author: Yeah, basically the entire difference as far as AS4 is returned between server-y and serverless integrations is that server-y integrations have good startup error handling (you first try to start the server machinery, and only if it works do you move on to actually listening for HTTP requests) and server-less integrations have crappy startup error handling (there's no way to do async work before accepting requests, so any startup errors just turn into "send a 500 for every request").

The proper pattern for server-y integrations involves asserting when you call a function like `expressMiddleware` that the server is started, so the startup issue should be a serverless integration or a poorly written server-y integration.

Reviewer: So this would be something more like this:
```suggestion
-  If Apollo Server hasn't correctly started up or is in the process of shutting down, it responds with a 500 status code. 
  - The former can happen if you use a serverless integration and it sends requests to an Apollo Server instance that hasn't properly started. The latter happens if you aren't properly [draining your server](/apollo-server/api/plugin/drain-http-server/#using-the-plugin).
```

Author: Oh the other thing is that serverless will wait to start... so it really means that there was an error. This should be good, I'll commit it:

So this would be something more like this:
```suggestion
-  If Apollo Server hasn't correctly started up or is in the process of shutting down, it responds with a 500 status code. 
  - The former can happen if you use a serverless integration and it sends requests to an Apollo Server instance that had an error on startup. The latter happens if you aren't properly [draining your server](/apollo-server/api/plugin/drain-http-server/#using-the-plugin).
```

> - If Apollo Server cannot parse the request into a legal GraphQL document and validate it against your schema, use status code 400. This also includes other problems with the request, such as attempting to send a batched HTTP request when `allowBatchedHttpRequests` is not enabled, or if CSRF prevention blocks the request.

Reviewer: ```suggestion
- If Apollo Server can't parse the request into a legal GraphQL document and validate it against your schema, it responds with a 400 status code. This can also happen with other request problems, such as if a client attempts to send a batched HTTP request when `allowBatchedHttpRequests` isn't enabled or if CSRF prevention blocks a request.
```

> - If an invalid HTTP method is used (`GET` with a mutation, or any HTTP method other than `GET` or `POST`), use status code 405.

Reviewer: ```suggestion
- If a request uses an invalid HTTP method (`GET` with a mutation, or any HTTP method other than `GET` or `POST`), then Apollo Server responds with a 405 status code.
```

> - If your `context` function throws, use status code 500.

Reviewer: ```suggestion
- If your `context` function throws, Apollo Server responds with a 500 status code.
```

> - If there is an unexpected error during the processing of the request (either a bug in Apollo Server, or a plugin hook that throws), use status code 500.

Reviewer: ```suggestion
- If there is an unexpected error during the processing of the request (either a bug in Apollo Server or a plugin hook throws), Apollo Server responds with a 500 status code.
```

> - Otherwise, use status code 200. This is essentially the case where the server is able to execute the GraphQL operation, and execution completes successfully (possibly with some resolver-specific errors).

Reviewer: ```suggestion
- Otherwise, Apollo Server returns a 200 status code. This is essentially the case where the server can execute the GraphQL operation, and execution completes successfully (though this can still include resolver-specific errors).
```

> You can affect the HTTP status code or set response headers either by throwing an error in a resolver or `context` function, or by writing a [plugin](/apollo-server/integrations/plugins). Be aware that GraphQL client libraries may not treat all response status codes the same, and so it will be up to your team to decide what patterns to use. We strongly encourage you to avoid setting the HTTP status code based on an error thrown from a resolver; setting the status code from context creation errors can be more useful.

Reviewer: How would you feel about breaking this out more? Something like this:
```suggestion
There are three ways to change an HTTP status code or set custom response headers. You can: throw an error in a resolver, throw an error in your `context` function, or you can write a [plugin](/apollo-server/integrations/plugins).

Be aware that GraphQL client libraries might not treat all response status codes the same, so it will be up to your team to decide which patterns to use. You should _avoid_ setting HTTP status codes based on errors thrown by resolvers. Instead, we recommend setting the HTTP status code based on `context` creation errors. 
```

Author: First paragraph: can the three options be written more parallel?

I like the second paragraph but maybe even write something like "While Apollo Server does allow you to set, HTTP status codes based on errors thrown by resolvers, best practices for GraphQL over HTTP encourage sending 200 whenever the operation executes" or something?

Reviewer: I'm sorry, I'm not sure what you mean by more parallel for the first paragraph. Like all in one sentence? 

How would you feel about this in terms of order:
```suggestion
There are three ways to change an HTTP status code or set custom response headers, you can: throw an error in a resolver, throw an error in your `context` function, or write a [plugin](/apollo-server/integrations/plugins).  

While Apollo Server does enable you to set HTTP status codes based on errors thrown by resolvers, best practices for GraphQL over HTTP encourage sending 200 whenever an operation executes.  Instead, we recommend setting the HTTP status code based on `context` creation errors. 

Additionally, be aware that GraphQL client libraries might not treat all response status codes the same, so it will be up to your team to decide which patterns to use. 
```

Author: I just meant that you reiterated "you can" in the third option, which you fixed even if you didn't understand me :)

Author: ```suggestion
There are three ways to change an HTTP status code or set custom response headers, you can: throw an error in a resolver, throw an error in your `context` function, or write a [plugin](/apollo-server/integrations/plugins).

While Apollo Server does enable you to set HTTP status codes based on errors thrown by resolvers, best practices for GraphQL over HTTP encourage sending 200 whenever an operation executes, so we don't recommend you use this mechanism in resolvers, just in the `context` function or in a plugin hooking into an early stage of the request pipeline.

Be aware that GraphQL client libraries might not treat all response status codes the same, so it will be up to your team to decide which patterns to use. 
```

Author: What do you think about this wording? I don't want to actively say "you should set the HTTP status code", like, people should do it if they have a reason to but it's fine not to. The note is just encouraging folks to *not* do that in certain cases

Reviewer: I love the first two parts; the middle sentence feels a bit long. I think splitting the middle sentence into two makes this flow just right:

```suggestion
There are three ways to change an HTTP status code or set custom response headers, you can: throw an error in a resolver, throw an error in your `context` function, or write a [plugin](/apollo-server/integrations/plugins).

While Apollo Server does enable you to set HTTP status codes based on errors thrown by resolvers, best practices for GraphQL over HTTP encourage sending 200 whenever an operation executes. So, we don't recommend using this mechanism in resolvers, just in the `context` function or in a plugin hooking into an early stage of the request pipeline.

Be aware that GraphQL client libraries might not treat all response status codes the same, so it will be up to your team to decide which patterns to use.
```

> To affect the HTTP status code and response headers from an error throw in a resolver or context function, throw a `GraphQLError` with an `http` extension:

Reviewer: ```suggestion
To change the HTTP status code and response headers based on an error thrown in either a resolver or `context` function, throw a `GraphQLError` with an `http` extension, like so:
```

<MultiCodeBlock>

```ts
import { GraphQLError } from 'graphql';

const resolvers = {
  Query: {
    someField() {
      throw new GraphQLError('the error message', {
        extensions: {
          code: 'SOMETHING_BAD_HAPPENED',
          http: {
            status: 404,
            headers: new Map([
              ['some-header', 'it was bad'],
              ['another-header', 'seriously'],
            ]),
          },
        },
      });
    }
  }
}
```

</MultiCodeBlock>

> You do not need to include `status` unless you want to override the default (200 for a resolver, 500 for a `context` function). You do not need to include `headers` unless you want to set headers. You should not rely on the behavior of what happens if multiple resolvers throw errors which set the status or set the same header; if you need to handle this case, write a plugin instead. (However, different resolver errors can set different headers.) Headers should be provided in a `Map` with lowercase names.

Reviewer: ```suggestion
You don't need to include `status` unless you want to override the default status code (200 for a resolver or 500 for a `context` function).   The optional `headers` field should provide a `Map` with lowercase header names.

If your setup includes multiple resolvers throwing errors that set the same status code or header, we recommend writing a plugin instead (as shown below).
```

Author: I would like to be explicit that this behavior is 'undefined' in the sense that we could change it in a release without it counting as backwards-incompatible.

Reviewer: How does something like this feel instead (being more clear this isn't a consistent thing):
```suggestion
You don't need to include `status` unless you want to override the default status code (200 for a resolver or 500 for a `context` function). The optional `headers` field should provide a `Map` with lowercase header names.

If your setup includes multiple resolvers which throw errors that set status codes or set the same header, the errors thrown are not guaranteed to be consistent. Instead, we recommend writing a plugin (as shown below).
```

Author: How about 

```suggestion
You don't need to include `status` unless you want to override the default status code (200 for a resolver or 500 for a `context` function). The optional `headers` field should provide a `Map` with lowercase header names.

If your setup includes multiple resolvers which throw errors that set status codes or set the same header, Apollo Server's may resolve this conflict in an arbitrary way which may change in future versions. Instead, we recommend writing a plugin (as shown below).
```

Reviewer: Small suggested tweaks for styling, but otherwise feels great:
```suggestion
You don't need to include `status` unless you want to override the default status code (200 for a resolver or 500 for a `context` function). The optional `headers` field should provide a `Map` with lowercase header names.

If your setup includes multiple resolvers which throw errors that set status codes or set the same header, Apollo Server might resolve this conflict in an arbitrary way (which could change in future versions). Instead, we recommend writing a plugin (as shown below).
```

> Alternatively, you can set the HTTP status code and headers from a plugin. As an example, here is how you could set a custom response header and status code based on a GraphQL error:

Reviewer: Changing the first sentence of this to reflect my suggestion above:
```suggestion
You can also set the HTTP status code and headers from a plugin. As an example, here is how you could set a custom response header and status code based on a GraphQL error:
```

- If there is an unexpected error during the processing of the request (either a bug in Apollo Server or a plugin hook throws), Apollo Server responds with a 500 status code.
- Otherwise, Apollo Server returns a 200 status code. This is essentially the case where the server can execute the GraphQL operation, and execution completes successfully (though this can still include resolver-specific errors).

There are three ways to change an HTTP status code or set custom response headers, you can: throw an error in a resolver, throw an error in your `context` function, or write a [plugin](/apollo-server/integrations/plugins).

While Apollo Server does enable you to set HTTP status codes based on errors thrown by resolvers, best practices for GraphQL over HTTP encourage sending 200 whenever an operation executes. So, we don't recommend using this mechanism in resolvers, just in the `context` function or in a plugin hooking into an early stage of the request pipeline.

Be aware that GraphQL client libraries might not treat all response status codes the same, so it will be up to your team to decide which patterns to use.

> To change the HTTP status code and response headers based on an error thrown in either a resolver or `context` function, throw a `GraphQLError` with an `http` extension, like so:

Reviewer 2: @glasser Maybe we should present the resolver example after the plugin example if it's not our official recommendation?
